source/WebApp/tests/component.codemirror.mode-asm-instructions.tests.ts

```diff
@@ -0,0 +1,42 @@
+import CodeMirror from 'codemirror';
+import '../components/internal/codemirror/mode-asm';
+
+test('mode-asm-instructions codemirror highlights correctly', async () => {
+
+    document.createRange = () => {
+        const range = new Range();
+
+        range.getBoundingClientRect = jest.fn();
+        range.getClientRects = () => ({
+            item: () => null,
+            length: 0,
+            [Symbol.iterator]: jest.fn()
+        });
+
+        return range;
+    };
+
+    const textarea = document.createElement('textarea');
+    document.body.appendChild(textarea);
+    const cm = CodeMirror.fromTextArea(textarea, {
+        readOnly: true,
+        mode: 'text/x-asm'
+    });
+
+    const assertHighlightedTextFor = (token: string, code: string) => {
+        const renderingListener = (instance: CodeMirror.Editor, line: CodeMirror.LineHandle, element: HTMLElement) =>
+            expect(element.innerHTML).toContain(`<span class="cm-builtin">${token}</span>`);
+
+        cm.on('renderLine', renderingListener);
+        cm.setValue(code);
+        cm.off('renderLine', renderingListener);

```

Comment:
There is a small issue here -- if there is a bug where CodeMirror never calls `renderLine`, the test will still pass.
Adding a suggestion based on `jest.fn()` (that also removes some typing and need for `off`).

Could not find a good matcher alternative to the loop though.

```suggestion
        const renderingListener = jest.fn();

        cm.on('renderLine', renderingListener);
        cm.setValue(code);

        expect(renderingListener).toBeCalled();
        for (const [,, element] of renderingListener.mock.calls) {
            expect((element as HTMLElement).innerHTML).toContain(`<span class="cm-builtin">${token}</span>`);
        }
```